File: src/robot_control/controllers/implementation/task_space_controller.py
# Created by Giuseppe Rizzi
# email grizzi@mavt.ehtz.ch

import numpy as np
import pinocchio as pin
from scipy.spatial.transform import Rotation

from robot_control.modeling import RobotWrapper
from robot_control.controllers import RobotControllerBase
from robot_control.controllers.utilities import get_nullspace


class TaskSpaceController(RobotControllerBase):
    def __init__(self, robot, controlled_frame):
        super(TaskSpaceController, self).__init__(robot)

        assert isinstance(robot, RobotWrapper)
        self.robot = robot
        self.controlled_frame = controlled_frame

        # Target is assumed to be in a world frame
        self.task_target = pin.SE3(np.eye(3), np.zeros((3, 1)))
        self.task_target_set = False

        # Task space gains
        self.kp = np.diag([10.0, 10.0, 10.0, 10.0, 10.0, 10.0])
        self.kd = np.diag([10, 10, 10, 10, 10, 10])

        # null space variables
        self.kqd_ns = np.diag([0.01] * self.robot.get_dof())
        self.kqp_res = np.diag([0.01] * self.robot.get_dof())
        self.q_rest = self.robot.get_neutral_configuration()

    def set_kp(self, kp):
        self.kp = kp

    def set_kd(self, kd):
        self.kd = kd

    def set_task_target(self, task_target):
        assert isinstance(task_target, pin.SE3)
        self.task_target = task_target
        self.task_target_set = True

    def compute_command(self):
        pose_current = self.robot.get_frame_placement(self.controlled_frame)
        pose_des = self.task_target
        if not self.task_target_set:
            pose_des = pose_current

        # pose error (local frame)
        dR = Rotation.from_matrix(pose_current.rotation.transpose().dot(pose_des.rotation)).as_quat()
        rotation_err = 2 * dR[3] * dR[:3]
        position_err = pose_current.rotation.transpose().dot(pose_des.translation - pose_current.translation)
        p_err = np.hstack([position_err, rotation_err])

        # velocity error (local frame)
        v_des = pin.log(pose_current.actInv(pose_des)).vector
        v_current = self.robot.get_frame_velocity(self.controlled_frame, pin.ReferenceFrame.LOCAL)
        v_err = v_des - v_current

        j, dj = self.robot.get_all_frame_jacobians(self.controlled_frame)
        self.robot.compute_all_terms()

        y = np.linalg.pinv(j, rcond=0.01).dot(self.kp.dot(p_err) + self.kd.dot(v_err) - dj.dot(self.robot.get_v()))
        eps = -self.kqd_ns.dot(self.robot.get_v()) - self.kqp_res.dot(self.robot.get_q() - self.q_rest)
        tau_null_space = get_nullspace(j).dot(eps)
        tau = self.robot.get_inertia().dot(y) + self.robot.get_nonlinear_terms() + tau_null_space
        return tau

File: robot_control/src/controllers/end_effector_controllers/task_space_controller.cc
#include "robot_control/controllers/end_effector_controllers/task_space_controller.h"

#include <pinocchio/spatial/se3.hpp>
#include <pinocchio/spatial/explog.hpp>

namespace pin = pinocchio;

namespace rc {

TaskSpaceController::TaskSpaceController(std::shared_ptr<RobotWrapper> wrp, std::string& controlled_frame) : controlled_frame(controlled_frame),
  svd(6, wrp->getDof()),
  J(6, wrp->getDof()),
  dJ(6, wrp->getDof()),
  solver(svd)
{
  wrapper = wrp;
  kp_ = Eigen::MatrixXd::Identity(6, 6) * 10.0;
  kd_ = Eigen::MatrixXd::Identity(6, 6) * 5.0;
  int dof = wrapper->getDof();
  kqd_ns = 0.1 * Eigen::MatrixXd::Identity(dof, dof);
  kqp_res = 0.1 * Eigen::MatrixXd::Identity(dof, dof);
  q_rest = wrapper->getNeutralConfiguration();
  target = pin::SE3(Eigen::Matrix<double, 3, 3>::Identity(), Eigen::Matrix<double, 3, 1>::Zero());
}

void TaskSpaceController::setTaskTarget(pin::SE3 task_target) {
  std::lock_guard<std::mutex> lock(target_mutex);
  target = task_target;
  target_set = true;
}

void TaskSpaceController::setKp(const Eigen::Matrix<double, 6, 1>& kp){ kp_ = kp.asDiagonal(); }

void TaskSpaceController::setKd(const Eigen::Matrix<double, 6, 1>& kd){ kd_ = kd.asDiagonal(); }

Eigen::VectorXd TaskSpaceController::computeCommand() {
  std::lock_guard<std::mutex> lock(target_mutex);
  pin::SE3& current_pose = wrapper->getFramePlacement(controlled_frame);
  pin::SE3 desired_pose;
  if (target_set) {
    desired_pose = target;
  } else {
    desired_pose = current_pose;
  }

  Eigen::Quaternion<double> dR(current_pose.rotation().transpose() * desired_pose.rotation());
  dR = dR.normalized();
  Eigen::Matrix<double, 6, 1> position_error;
  position_error.head<3>() = current_pose.rotation().transpose() * (desired_pose.translation() - current_pose.translation());
  position_error[3] = 2.0 * dR.w() * dR.x();
  position_error[4] = 2.0 * dR.w() * dR.y();
  position_error[5] = 2.0 * dR.w() * dR.z();

  pin::Motion velocity = pin::log6(current_pose.actInv(desired_pose));
  Eigen::Matrix<double, 6, 1> desired_velocity;
  desired_velocity.head<3>() = velocity.linear();
  desired_velocity.tail<3>() = velocity.angular();

  pin::Motion frame_velocity = wrapper->getFrameVelocity(controlled_frame);
  Eigen::Matrix<double, 6, 1> current_velocity;
  current_velocity.head<3>() = frame_velocity.linear();
  current_velocity.tail<3>() = frame_velocity.angular();
  Eigen::Matrix<double, 6, 1> velocity_error = desired_velocity - current_velocity;

  wrapper->getAllFrameJacobians(controlled_frame, J, dJ);
  wrapper->computeAllTerms();

  // task space dynamics
  Eigen::VectorXd error = kp_ * position_error + kd_ * velocity_error - dJ * wrapper->v;
  solver.compute(J);
  Eigen::VectorXd y = solver.solve(error);
  return wrapper->getInertia() * y  + wrapper->getNonLinearTerms();
}

Eigen::VectorXd TaskSpaceController::advance(Eigen::VectorXd& q, Eigen::VectorXd& v) {
  wrapper->updateState(q, v, true);
  return computeCommand();
}

}
